Source/Provers/SMTLib/SMTLibProcess.cs

```diff
@@ -466,35 +431,39 @@ void DisposeProver()
 
     void prover_OutputDataReceived(object sender, DataReceivedEventArgs e)
     {
-      lock (this)
-      {
-        if (e.Data != null)
+        if (e.Data == null)
         {
-          if (options.Verbosity >= 2 || (options.Verbosity >= 1 && !e.Data.StartsWith("(:name ")))
-          {
-            Console.WriteLine("[SMT-OUT-{0}] {1}", smtProcessId, e.Data);
-          }
+          return;
+        }
+
+        if (options.Verbosity >= 2 || (options.Verbosity >= 1 && !e.Data.StartsWith("(:name ")))
+        {
+          Console.WriteLine("[SMT-OUT-{0}] {1}", smtProcessId, e.Data);
+        }
 
-          proverOutput.Enqueue(e.Data);
-          Monitor.Pulse(this);
+        lock (this) {
+          if (outputReceivers.TryDequeue(out var source)) {
+            source.SetResult(e.Data);
+          } else {
+            proverOutput.Enqueue(e.Data);
+          }
         }
-      }
     }
 
     void prover_ErrorDataReceived(object sender, DataReceivedEventArgs e)
     {
-      lock (this)
-      {
-        if (e.Data != null)
+      lock (this) {
+        if (e.Data == null)
```

Comment:
In the method above you moved the e.data test out of the lock, but not here.  Could you do the same here?  It would avoid taking the lock when `e.Data` is `null`